File: Python/Leetcode/hashTable/lcci0102.check-permutation-lcci.py
from self import self


# 哈希表
def CheckPermutation(self, s1: str, s2: str) -> bool:
    hmap = dict()
    for i in s1:
        if i not in hmap:
            hmap[i] = 1
            continue
        hmap[i] += 1

    for i in s2:
        if i not in hmap:
            return False
        hmap[i] -= 1

    for i in hmap.values():
        if i != 0:
            return False

    return True


# 偷鸡做法，排序
def CheckPermutationSort(self, s1: str, s2: str) -> bool:
    return sorted(s1) == sorted(s2)


print(CheckPermutation(self, "asvnpzurz", "urzsapzvn"))

File: Cpp/Leetcode/hashTable/lcci0102.check-permutation-lcci.cpp
#include "LC.h"

// 哈希表做法
bool CheckPermutation(string s1, string s2) {
    if(s1.size() != s2.size()) return false;
    unordered_map<char, int> umap;
    for(auto i : s1) {
        ++umap[i];
    }
    for(auto i : s2) {
        --umap[i];
    }
    for(auto i : umap) {
        if(i.second != 0) {
            return false;
        }
    }
    return true;
}

// 调库偷鸡做法
bool CheckPermutationSort(string s1, string s2) {
    if(s1.size() != s2.size()) return false;
    sort(s1.begin(), s1.end());
    sort(s2.begin(), s2.end());
    for(int i = 0; i < s1.size(); ++i) {
        if(s1[i] != s2[i]) return false;
    }
    return true;
}

int main () {
    string s1 = "asvnpzurz", s2 = "urzsapzvn";
    cout << boolalpha << CheckPermutation(s1, s2);
    return 0;
}